subsys/infinispan/src/main/java/org/commonjava/indy/subsys/infinispan/CacheProducer.java

```diff
@@ -200,4 +186,79 @@ public String getId()
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( "Cannot read infinispan configuration from : " + path, e );
+        }
+
+        StringSearchInterpolator interpolator = new StringSearchInterpolator();
+        interpolator.addValueSource( new PropertiesBasedValueSource( System.getProperties() ) );
+
+        try
+        {
+            configuration = interpolator.interpolate( configuration );
+        }
+        catch ( InterpolationException e )
+        {
+            throw new RuntimeException( "Cannot resolve expressions in infinispan configuration from: " + path, e );
+        }
+        return configuration;
+    }
+
+    /**
+     * For the ISPN merging, we should involve at least two different xml config scopes here,
+     * one is from indy self default resource xml, another one is from customer's config xml.
+     *
+     * To prevent the error of EmbeddedCacheManager instances configured with same JMX domain,
+     * ISPN should enable 'allowDuplicateDomains' attribute for per GlobalConfigurationBuilder build,
+     * that will cost more price there for DefaultCacheManager construct and ConfigurationBuilder build.
+     *
+     * Since what we need here is simply parsing xml inputStreams to the defined configurations that ISPN
+     * could accept, then merging the two stream branches into a entire one.
+     * What classes this method uses from ISPN are:
+     * {@link ConfigurationBuilderHolder}
+     * {@link ParserRegistry}
+     * {@link ConfigurationManager}
+     *
+     * @param resourceStr
+     * @param confStr
+     * @return
+     * @throws IOException
+     */
+    private EmbeddedCacheManager mergeIspnConfiguration( String resourceStr, String confStr )
+    {
+        EmbeddedCacheManager mergedManager = new DefaultCacheManager();
+
+        ConfigurationBuilderHolder confHolder = ( new ParserRegistry() ).parse( IOUtils.toInputStream( confStr ) );
+        ConfigurationManager confManager = new ConfigurationManager( confHolder );
+
+        for ( String name : confManager.getDefinedCaches() )
+        {
+            logger.info( "[ISPN merging] Define cache {} from customer config.", name );
+            mergedManager.defineConfiguration( name, confManager.getConfiguration( name ) );
+        }
+
+        ConfigurationBuilderHolder resourceHolder =
+                ( new ParserRegistry() ).parse( IOUtils.toInputStream( resourceStr ) );
+        ConfigurationManager resourceManager = new ConfigurationManager( resourceHolder );
+
+        for ( String name : resourceManager.getDefinedCaches() )
+        {
+            if ( !mergedManager.getCacheNames().contains( name ) )
+            {
+                logger.info( "[ISPN merging] Define cache {} from classpath resource config.", name );
+                mergedManager.defineConfiguration( name, resourceManager.getConfiguration( name ) );
+            }
+        }
+
+        return mergedManager;
```

Comment:
Are there any configurations that might not be contained in a defined cache, which we should be migrating from one or both of these infinispan config files?

Reply:
I think ISPN should declare the caches in two: named caches, and default cache. Default cache is named as ‘___defaultcache’, for the defined/named caches, whenever they are specified, named caches inherit settings from the default cache while additional behavior can be specified or overridden, this is what I found in the official doc.

I try to use the EmbeddedCacheManager api to define/set in the default cache, it’s not allowed, it will say: ISPN000370: Cache name '___defaultcache' cannot be used as it is a reserved, internal name. So ISPN should only allow define the configuration for named caches.

And When I go through the official doc, http://docs.jboss.org/infinispan/8.2/configdocs/infinispan-config-8.2.html , it’s specified the default-cache is defined in the <cache-container> element of the XML configuration file, while this kind of cache must be named and rightly included in the scope of defined/named caches (like indy’s ‘local’ cache).

So I suppose the ‘___defaultcache’ cache and its configuration is a base one for the named caches, should distinguish from the ‘default-cache’ we defined in xml schema. All the caches which allow user to specify in xml schema should be overridden from the inner base cache - ‘___defaultcache’ and also be named.

So from what I know, the defined cache should cover all of indy’s current cases, @jdcasey , I’m not sure what I know is absolutely right, it’s a short time for me to get deeper on ISPN, if any need to be correct, pls let me know.

Reply:
You should not use the default cache. We killed it in 9.x, transforming it to a named cache.
Also the code above could be replaced with a loop with two iterations (one per file).

Reply:
@tristantarrant Thanks a lot for your review, and it's reassuring that we don't handle the default cache opertaions here.